barretenberg/cpp/src/barretenberg/stdlib/hash/poseidon2/poseidon2_permutation.hpp

```diff
@@ -0,0 +1,212 @@
+            NativePermutation::add_round_constants(current_native_state, round_constants[i]);
+            NativePermutation::apply_sbox(current_native_state);
+            NativePermutation::matrix_multiplication_external(current_native_state);
+            for (size_t j = 0; j < t; ++j) {
+                current_state[j] = witness_t<Builder>(builder, current_native_state[j]);
+            }
+        }
+
+        const size_t p_end = rounds_f_beginning + rounds_p;
+        for (size_t i = rounds_f_beginning; i < p_end; ++i) {
+            poseidon2_internal_gate_<FF> in{ current_state[0].witness_index,
+                                             current_state[1].witness_index,
+                                             current_state[2].witness_index,
+                                             current_state[3].witness_index,
+                                             i };
+            builder->create_poseidon2_internal_gate(in);
+            current_native_state[0] += round_constants[i][0];
+            NativePermutation::apply_single_sbox(current_native_state[0]);
+            NativePermutation::matrix_multiplication_internal(current_native_state);
+            for (size_t j = 0; j < t; ++j) {
+                current_state[j] = witness_t<Builder>(builder, current_native_state[j]);
+            }
+        }
+
+        for (size_t i = p_end; i < NUM_ROUNDS; ++i) {
+            poseidon2_external_gate_<FF> in{ current_state[0].witness_index,
+                                             current_state[1].witness_index,
+                                             current_state[2].witness_index,
+                                             current_state[3].witness_index,
+                                             i };
+            builder->create_poseidon2_external_gate(in);
+            // calculate the new witnesses
+            NativePermutation::add_round_constants(current_native_state, round_constants[i]);
+            NativePermutation::apply_sbox(current_native_state);
+            NativePermutation::matrix_multiplication_external(current_native_state);
+            for (size_t j = 0; j < t; ++j) {
+                current_state[j] = witness_t<Builder>(builder, current_native_state[j]);
+            }
+        }
+        // need to add an extra row here to ensure that things check out
+        poseidon2_end_gate_<FF> in{
+            current_state[0].witness_index,
+            current_state[1].witness_index,
+            current_state[2].witness_index,
+            current_state[3].witness_index,
+        };
+        builder->create_poseidon2_end_gate(in);
+        return current_state;
+    }
+
+    static void initial_external_matrix_multiplication(Builder* builder, State& state)
+    {
+        // create the 6 gates for the initial matrix multiplication
+        // gate 1: Compute tmp1 = state[0] + state[1] + 2 * state[3]
+        FF tmp1 = state[0].get_value() + state[1].get_value() + FF(2) * state[3].get_value();
```

Comment:
might be doing things weirdly here

Reply:
use witness_t where possible

Reply:
updated, but maybe double check the usage, since it seems weird